### PR TITLE
Upgrade tested Python and Django versions

### PR DESCRIPTION
**Pull request by adamchainz**
This PR:

1. Drops Python 3.7, since it is EOL.
2. Adds Python 3.11 and 3.12.
3. Drops Django 3.2. The Ubuntu CI environment uses SpatiaLite 5.0, but that is only supported by Django 4.0+, since [Ticket #32575](https://code.djangoproject.com/ticket/32575). It’s easiest to drop Django 3.2, which ends its long term support [this month](https://www.djangoproject.com/download/#supported-versions) anyway.
4. Adds Django 4.2 and 5.0. No code changes required.
5. Changes GitHub actions to run one task per Python version, running all Django versions underneath. That is faster and removes the need to duplicate test matrices between tox and GHA.
6. Upgrades GitHub Actions actions repositories.
7. Re-applies isort, which makes some changes from [version 5](https://pycqa.github.io/isort/docs/major_releases/introducing_isort_5.html), mostly due to working on all kinds of import blocks.
8. Adds a workaround for a bug when running SQLite 3.36+ with SpatiaLite 5 - see comment in `quicktest.py`.
9. Adjusts a test to pass with slightly altered values - I think caused by SpatiaLite 5, which [changed its topology library](https://groups.google.com/g/spatialite-users/c/MlXsPU0ZMYw/m/t-ynEJ2TBQAJ), among other things that may change calculations.

### CODE REVIEW

**Comment by adamchainz** [comment]
Alright, ready for review.

@smithdc1 : If you have any time to check the workaround in `quicktest.py`, that would be appreciated. Your name appeared a lot whilst I was working on this :)

**Comment by smithdc1** [comment]
Hi Adam 👋  -- yes happy to have a look at this in this morning. 

Just one initial comment, it seems that the `install_requires` and `python_requires` in `setup.py` need a version bump too.

File: djgeojson/tests.py [hunk 1]
```diff
@@ -680,9 +679,9 @@ def test_within_viewport(self):
         response = self.view.render_to_response(context={})
         geojson = json.loads(smart_str(response.content))
         self.assertEqual(len(geojson['features']), 2)
-        self.assertAlmostEqual(geojson['features'][0]['geometry']['coordinates'][0], 6.843322039261242)
+        self.assertAlmostEqual(geojson['features'][0]['geometry']['coordinates'][0], 6.843321961076886)
         self.assertAlmostEqual(geojson['features'][0]['geometry']['coordinates'][1], 52.76181518632031)
-        self.assertAlmostEqual(geojson['features'][1]['geometry']['coordinates'][0], 6.846053318324978)
+        self.assertAlmostEqual(geojson['features'][1]['geometry']['coordinates'][0], 6.846053240233331 )
```
Comment (smithdc1) [inline djgeojson/tests.py:684]:
```suggestion
        self.assertAlmostEqual(geojson['features'][1]['geometry']['coordinates'][0], 6.846053240233331)
```

File: quicktest.py [hunk 2]
```diff
@@ -69,6 +69,22 @@ def run_tests(self):
 
         django.setup()
 
+        # Workaround for incompatibility between SQLite 3.36+ and SpatiaLite 5,
```
Comment (smithdc1) [inline quicktest.py:72]:
```suggestion
        # Workaround for incompatibility between SQLite 3.36+ and SpatiaLite 5.0,
```

I've not tested many version combinations but it seems that SpatiaLite 5.0 is the issue. The tests pass for me with SpatiaLite 5.1 and SQLite 3.45.2 (appreciate it is also a much more modern version of SQLite)

File: djgeojson/tests.py (hunk 1 above)
Comment (adamchainz) [reply djgeojson/tests.py]: Very eagle-eyed! Could do with Black here really...

**Comment by adamchainz** [comment]
> Just one initial comment, it seems that the `install_requires` and `python_requires` in `setup.py` need a version bump too.

Thanks, done.

**Review by smithdc1 — APPROVED** [review]
Hi Adam

This looks good to me. It's interesting to see how this issue can be worked around!

It seems that SpatiaLite 5.1 fixed this issue so when that's more easily available the workaround can be removed. I'm guessing it's not worth the effort to try and install that in place of workaround in `quicktest.py` now that this is working nicely.

**Review by smithdc1 — APPROVED** [review]

File: quicktest.py (hunk 2 above)
Comment (adamchainz) [reply quicktest.py:72]: Thank you for testing. Yeah, I guess to update SpatiaLite, we would need to update the Ubuntu version on GitHub Actions, when 24.04 becomes available.

Are there any clear release notes for SpatiaLite 5.1? I found the site hard to navigate.

Comment (smithdc1) [reply quicktest.py:72]: I couldn't find any release notes either.

**Comment by Gagaro** [comment]
Thanks for the PR :+1: . I'll merge as soon as the CI finishes.

**Pull request merged**